tests/test_slips_utils.py

```diff
@@ -1,10 +1,318 @@
 from tests.module_factory import ModuleFactory
+import datetime
+import ipaddress
+from unittest.mock import patch
+import requests
+from dataclasses import dataclass
+from enum import Enum
+import sys
+import pytest
 
 
 def test_get_hash_from_file():
     utils = ModuleFactory().create_utils_obj()
     # a file that we know doesn't change
     assert (
-        utils.get_hash_from_file("modules/template/__init__.py")
-        == "2d12747a3369505a4d3b722a0422f8ffc8af5514355cdb0eb18178ea7071b8d0"
+        utils.get_hash_from_file('modules/template/__init__.py')
+        == '879d569533ed70a30c2a2e77fe5ae14d5a30606df470f1c354685ee40471140b')
+    
+def test_get_hash_from_nonexistent_file():
+    utils = ModuleFactory().create_utils_obj()
+    with pytest.raises(FileNotFoundError):
+        utils.get_hash_from_file('nonexistent_file.txt')
+
+def test_get_hash_from_file_permission_error():
+    utils = ModuleFactory().create_utils_obj()
+    with patch('builtins.open', side_effect=PermissionError):
+        with pytest.raises(PermissionError):
+            utils.get_hash_from_file('restricted_file.txt')    
+
+def test_sanitize():
+    utils = ModuleFactory().create_utils_obj()
+    input_string = "Hello; world `& |$(this"
+    expected_output = "Hello world ` this"
+    assert utils.sanitize(input_string) == expected_output
+    input_string = "This is a clean string"
+    assert utils.sanitize(input_string) == input_string
+    
+def test_sanitize_with_all_special_characters():
+    utils = ModuleFactory().create_utils_obj()
+    special_chars = ";`&|$("
+    input_string = f"Hello{special_chars}world"
+    expected_output = "Hello`world"
+    assert utils.sanitize(input_string) == expected_output     
+    
+def test_detect_data_type():
+    utils = ModuleFactory().create_utils_obj()
+    assert utils.detect_data_type('192.168.1.100') == 'ip'
+    assert utils.detect_data_type('2001:0db8:85a3:0000:0000:8a2e:0370:7334') == 'ip'
+    assert utils.detect_data_type('192.168.0.0/16') == 'ip_range'
+    assert utils.detect_data_type('e10adc3949ba59abbe56e057f20f883e') == 'md5'
+    assert utils.detect_data_type('example.com') == 'domain'
+    assert utils.detect_data_type('http://example.com/some/path') == 'url'
+    assert utils.detect_data_type('AS12345') == 'asn'
```

Comment:
Hey @Sekhar-Kumar-Dash why not use? i think it will be cleaner
`@pytest.mark.parametrize("data, expected_type",...)`